### PR TITLE
feat(kubernetes_platform): Update kubernetes_platform go package to include node affinities and pod (anti)affinities

### PR DESCRIPTION
**Pull request by cjidboon94**
**Description of your changes:**
Part of #9768 and #9682 

This PR adds Node Affinities and Pod Affinities and Anti-Affinities to the kubernetes platform proto file.  It also regenerates the Golang stub.

To quote @droctothorpe mentioned in his PR
> This PR follows the piece-meal approach that @Tomcli led by example with in https://github.com/kubeflow/pipelines/pull/10357 and clarified in https://github.com/kubeflow/pipelines/pull/10357#issuecomment-1885881465.
> We will submit a follow up PR that updates driver.go and another PR that updates the SDK.

All the relevant information about node affinities and pod (anti)affinity is captured in the NodeAffinity and PodAffinity proto messages and should be enough to generate all possible node- and podaffinities.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: kubernetes_platform/proto/kubernetes_executor_config.proto [hunk 1]
```diff
@@ -176,3 +178,23 @@ message Toleration {
     string effect = 4;
     optional int64 toleration_seconds = 5;
 }
+
+message NodeAffinity{
+    string key = 1;
+    string operator = 2;
+    string value = 3;
+    //Setting the weight makes it use PreferredDuringSchedulingIgnoredDuringExecution type of affinities instead of RequiredDuringSchedulingIgnoredDuringExecution type of affinities
```
Comment (Tomcli) [inline kubernetes_platform/proto/kubernetes_executor_config.proto:186]: Just to confirm, so by default we are using `RequiredDuringSchedulingIgnoredDuringExecution` if weight is not defined?

File: kubernetes_platform/proto/kubernetes_executor_config.proto [hunk 2]
```diff
@@ -34,6 +34,8 @@ message KubernetesExecutorConfig {
     int64 active_deadline_seconds = 10;
     repeated FieldPathAsEnv field_path_as_env = 11;
     repeated Toleration tolerations = 12;
+    repeated NodeAffinity node_affinities = 13;
```
Comment (Tomcli) [inline kubernetes_platform/proto/kubernetes_executor_config.proto:37]:
```suggestion
    repeated NodeAffinity node_affinity = 13;
```

In Kubernetes spec, the node affinity key name is `nodeAffinity`. We should use plural name such as `tolerations` if it's defined the same way in Kubernetes spec.
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#nodeaffinity-v1-core

File: kubernetes_platform/proto/kubernetes_executor_config.proto [hunk 3]
```diff
@@ -34,6 +34,8 @@ message KubernetesExecutorConfig {
     int64 active_deadline_seconds = 10;
     repeated FieldPathAsEnv field_path_as_env = 11;
     repeated Toleration tolerations = 12;
+    repeated NodeAffinity node_affinities = 13;
+    repeated PodAffinity pod_affinities = 14;
```
Comment (Tomcli) [inline kubernetes_platform/proto/kubernetes_executor_config.proto:38]:
```suggestion
    repeated PodAffinity pod_affinity = 14;
```

**Review by Tomcli — CHANGES_REQUESTED** [review]

File: kubernetes_platform/proto/kubernetes_executor_config.proto (hunk 1 above)
Comment (cjidboon94) [reply kubernetes_platform/proto/kubernetes_executor_config.proto:186]: That's my intention indeed. if we do not set weight, weight default to a value of 0.  (See https://github.com/cjidboon94/pipelines/blob/c902efcef860b2e96a79693fa2336c18b66ae73c/kubernetes_platform/go/kubernetesplatform/kubernetes_executor_config.pb.go#L1198-L1203 )

However, `weight` field (which is only used in the `PreferredDuringSchedulingIgnoredDuringExecution` Node/PodAffinity) is required to be a value between 1 and 100. So, if weight is equal to 0 , that should indicate to the driver to set a `RequiredDuringSchedulingIgnoredDuringExecution`

I will add the necessary transforms in the driver.go PR and checks in the python PR, when this one is approved.

File: kubernetes_platform/proto/kubernetes_executor_config.proto (hunk 2 above)
Comment (cjidboon94) [reply kubernetes_platform/proto/kubernetes_executor_config.proto:37]: Fair enough. One thing, I am not following the Kubernetes spec, but intend to generate the `spec.affinity` section using the data from the list of NodeAffinity/PodAffinity object. 
If I should follow the k8s api spec more closely, let me know and will rethink how to set this up

Comment (Tomcli) [reply kubernetes_platform/proto/kubernetes_executor_config.proto:37]: You don't have to follow the Kubernetes api spec. I'm just thinking using the `node_affinity` name might make more sense in terms of SEO. The only other place I saw people use `node_affinities` is in GCP compute instance.

Comment (cjidboon94) [reply kubernetes_platform/proto/kubernetes_executor_config.proto:37]: Fair enough. Seems like all references are singular, indeed. Also from the definition in the spec:
> Affinity is a group of affinity scheduling rules.
So I've renamed the field in the config to pod_affinity and node_affinity as you have suggested. And the classnames to `NodeAffinityRule` and `PodAffinityRule`. Then `pod_affinity` is a list of `PodAffinityRule`s and `node_affinity` is a list of `NodeAffinityRule`s being consistent with the semantics (a group of rules)

**Review by Tomcli — COMMENTED** [review]
/lgtm
/assign @chensun

**Comment by cjidboon94** [comment]
I see there's still a few things missing in the proto. There's no way to distinguish between matchExpressions/matchLabels/matchFields (currently assumed it was all just matchExpressions). And there's some namespace selection stuff that are missing for PodAffinity. Willl add these tomorrow

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#podaffinityterm-v1-core
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#nodeselectorterm-v1-core

**Review by Tomcli — COMMENTED** [review]
/lgtm

**Review by chensun — APPROVED** [review]
/lgtm

**Comment by rimolive** [comment]
@cjidboon94 Can you please rebase this PR?

**Comment by cjidboon94** [comment]
> @cjidboon94 Can you please rebase this PR?

Done

**Pull request merged**